### ISSUE TITLE
Investigate removing jQuery dependency

### ISSUE THREAD

**Issue by danroth27** [aspnet/AspNetCore#8573]
Bootstrap 5 is removing it's dependency on jQuery: https://github.com/twbs/bootstrap/pull/23586

Should we do the same?

**Comment by mkArtakMSFT** [comment]
@ryanbrandenburg can you please investigate and come up with some cost for this work? Thanks!

**Comment by mkArtakMSFT** [comment]
@danroth27 seems we rely on jQuery for our validation too, it's not just transitive because of Bootstrap dependency.

**Comment by danroth27** [comment]
@mkArtakMSFT Right, the main work item tracked by this issue is investigating what it would take to remove the jQuery dependency from our validation libraries.

**Comment by danroth27** [comment]
Given that our current validation support is built on jQuery validation, the way I would think about this issue is to take a fresh look at what our client validation support should look like. Is there a different validation library that we should consider using? Etc.

**Comment by ryanbrandenburg** [comment]
The one option that seems to be most popular (and standards compliant) would be using the built-in HTML5 form validation. The problem there is that it would only work for client-side validation. As far as I can tell HTML5 form validation doesn't solve any server-side validation problems, and even if it did it would require a lot of changes to DataAnnotations. HTML5 also isn't supported on older browsers (IE9 was the example I found).

In my opinion it's not worth the work that would be required to remove our dependencies here in the absence of a more unified vision about what value we can produce from re-working DataAnnotations.

**Comment by Muchiachio** [comment]
We had the same problem and we actually wrote a vanilla js library for our project needs. It's mostly the same as jquery validate except some javascript api methods, but default asp net core template don't use them anyway.

If some else comes here looking for a solution, you can check out our [Wellidate](https://github.com/NonFactors/Wellidate) library. I recently made it open source after adding the missing functionality. It's intended to be a drop in replacement in asp net core and other projects using jquery validation. It also supports new bootstrap/customValidity based styling too.

I'm currently running through docs and use cases to even out rough edges, so if anyone wants to test it out before public release, feel free.

**Comment by jerriep** [comment]
+1 for this - especially since other jQuery-free CSS frameworks like Tailwind are also gaining popularity. For other people stumbling on this issue in the future, I came across this library that appears to be a good replacement in the meantime:

https://github.com/haacked/aspnet-client-validation

**Comment by xt0rted** [comment]
There's also https://github.com/NonFactors/Wellidate. I haven't used it but it's on my list of things to look into when I start removing jquery from my scaffolded identity pages.

**Comment by pranavkm** [comment]
Moving this out of 6.0. The initial thought was that we could do something tactical building upon either browser features such as constraint validation or hand rolling a facsimile of jQuery.validation. But on further inspection, we realize that an alternative that simply removes jquery does not move the needle enough for us invest in. 

https://github.com/aspnet/specs/blob/main/notes/RemoveJQueryUnobtrusiveValidation.md has some notes about our conversations around this and what we might expect if we were to consider improving this experience.

**Comment by Andrzej-W** [comment]
@pranavkm your link to https://github.com/aspnet/specs/blob/main/notes/RemoveJQueryUnobtrusiveValidation.md is invalid

**Comment by pranavkm** [comment]
Sorry @Andrzej-W that repo is non-public, but I've updated my comment with details about what we would want from a re-design.

**Comment by jerriep** [comment]
@pranavkm Thanks for posting those notes. I am eager to see what you can come up with in a future version

**Comment by javiercn** [comment]
https://github.com/dotnet/aspnetcore/issues/41750

**Comment by daikoz** [comment]
Build-in Client-side form validation:
https://developer.mozilla.org/en-US/docs/Learn/Forms/Form_validation

Bootstrap usage:
https://getbootstrap.com/docs/5.2/forms/validation/

**Comment by daikoz** [comment]
_There is no standard way to change their look and feel with CSS._
Of course, I'm agree. But if you check validitystate and update le class name in javascript on client side, it is a workaround.

_They depend on the browser locale, which means that you can have a page in one language but an error message displayed in another language._
If .net keep the data-* for message and manage it on client side to display it. It solve the issue.

**Comment by sanamhub** [comment]
This issue is still unplanned?! wow..

**Comment by sanamhub** [comment]
@danroth27

**Comment by dahlbyk** [comment]
It seems reasonable to continue iterating on https://github.com/haacked/aspnet-client-validation rather than reinvent a first-party equivalent. We still haven't shipped 1.0, so breaking API feedback is welcome. Need a nonzero testing story, too. Sponsors? 😁 

https://github.com/dotnet/aspnetcore/pull/34777 was a great solution to https://github.com/dotnet/aspnetcore/issues/17412 (other than deleting `public` members that should be marked `[Obsolete]` instead). The documentation for the new properties should include a warning like:

> Changing this value is incompatible with jQuery Unobtrusive Validation 4.x and lower.

...with a link to docs that can be updated to reflect the current state of the art. That may eventually be "it Just Works™️ with jUV 5.x or `aspnet-client-validation`" but in the meantime it can enumerate known limitations/options.

I don't think anyone is surprised that updating jUV isn't a top priority for the team, but waiting on a perfect solution has thus far meant there's no solution (except reflection). A well-documented optional configuration seems like a great first step that can be mostly community-driven.

**Comment by jburkett** [comment]
Can we at least get jQuery removed from the template? Encouraging use of jQuery should be considered an active bad, in and of itself, at this point. Experienced web developers will have no problem using web standards or finding a replacement package (say, https://github.com/haacked/aspnet-client-validation). Novice developers are hindered in their growth by promoting the use of libraries that have been outdated for the better part of a decade.

**Comment by sanamhub** [comment]
Any updates in this?

**Comment by Neutrino-Sunset** [comment]
There's a great tutorial on how to use aspnet-client-validation [here](https://andrewlock.net/adding-client-side-validation-to-aspnet-core-without-jquery-or-unobtrusive-validation/).

Can anyone explain the cause of all the navel gazing here? Doesn't addressing this consist of nothing more complicated than updating the ASP Core project template to use a newer and better library that already exists?

To me this comes across as the usual MS attitude of 'we aren't going to fix this because we are already working on something new' (Blazor?), which I think is the main reason why most developers choose to use frameworks written by anyone other than MS.

**Comment by andkorsh** [comment]
@Neutrino-Sunset especially for identity templates

**Comment by fredericDelaporte** [comment]
Replacing the current validation libraries by https://github.com/haacked/aspnet-client-validation would be a progress. Still, could we also have an option to not have client side validation enabled on tag helpers? It would allow to use whatever we like, instead of being forced into one library. (Especially if it does not handle localization and globalization, meaning, not supporting culture specific way of input numbers with decimals, by example.)

Sure, we can just not include the javascript validation libraries in the page, but that will still bloat the generated HTML with many useless attributes.

(My personal take would actually having tag helpers generating standard HTML validation attributes, free to the developer to add a layer of javascript to customize the client validation if need be.)

**Comment by dahlbyk** [comment]
> Some have mentioned https://github.com/haacked/aspnet-client-validation, but even that is dated - it's meant as a drop-in replacement. It does not improve the stack.

It improves the stack by removing a dependency on jQuery. 🤷‍♂️ Hard not to be dated when aiming for backwards compatibility.

> All modern frameworks use the [Constraint Validation API](https://caniuse.com/constraint-validation), which is considered part of "baseline", and available in all browsers (since 2018!). Yet we're still stuck in the web stone age.

_[citation needed]_

If all modern frameworks do this then I expect it would be easy to provide examples of how different frameworks automatically progressively enhance server-side validation with client-side validation using this API.

**Comment by grokky1** [comment]
I don't think it's a bad option - it's better than what we have now. *At the very least*, it would be good to replace jquery / etc. with that.

But it would be better to adopt modern standards. Check that link I posted - that API has been widely available since 2018.

Also I didn't mean that we can simply throw out what we have now and magically change to validation constraints. Some work must be done to integrate it with tag helpers and data annotations.

**Comment by Atulin** [comment]
So, it's been over 6 years now, and we're still in a limbo...?

**Comment by duranserkan** [comment]
I came here after reading [Andrew Lock's article.](https://andrewlock.net/adding-client-side-validation-to-aspnet-core-without-jquery-or-unobtrusive-validation/) I am amazed by the situation.

**Comment by oroztocil** [comment]
The current plan is to implement a new JS library (without the jQuery dependency) that would both cover https://github.com/dotnet/aspnetcore/issues/51040 and could be used as a drop-in replacement for MVC.

**Comment by ztrv** [comment]
Seems like given it’s been 7 years we can effectively give up on this ever happening

**Comment by Atulin** [comment]
Calling it now, the issue will eventually be closed with "uhhhhh use Blazor instead ig"
